cpp/include/raft/core/operators.hpp

```diff
@@ -0,0 +1,262 @@
+  template <typename T1, typename T2>
+  constexpr RAFT_INLINE_FUNCTION auto operator()(T1 a, T2 b) const
+  {
+    return a - b;
+  }
+};
+
+struct mul_op {
+  template <typename T1, typename T2>
+  constexpr RAFT_INLINE_FUNCTION auto operator()(T1 a, T2 b) const
+  {
+    return a * b;
+  }
+};
+
+struct div_op {
+  template <typename T1, typename T2>
+  constexpr RAFT_INLINE_FUNCTION auto operator()(T1 a, T2 b) const
+  {
+    return a / b;
+  }
+};
+
+struct div_checkzero_op {
+  template <typename Type>
+  constexpr RAFT_INLINE_FUNCTION Type operator()(Type a, Type b) const
+  {
+    if (b == Type{0}) { return Type{0}; }
+    return a / b;
+  }
+};
+
+struct pow_op {
+  template <typename Type>
+  constexpr RAFT_INLINE_FUNCTION Type operator()(Type a, Type b) const
+  {
+    return std::pow(a, b);
+  }
+};
+
+struct min_op {
+  template <typename Type>
+  constexpr RAFT_INLINE_FUNCTION Type operator()(Type a, Type b) const
+  {
+    if (a > b) { return b; }
+    return a;
+  }
+};
+
+struct max_op {
+  template <typename Type>
+  constexpr RAFT_INLINE_FUNCTION Type operator()(Type a, Type b) const
+  {
+    if (b > a) { return b; }
+    return a;
+  }
+};
+
+struct sqdiff_op {
+  template <typename Type>
+  constexpr RAFT_INLINE_FUNCTION Type operator()(Type a, Type b) const
+  {
+    Type diff = a - b;
+    return diff * diff;
+  }
+};
+
+struct argmin_op {
+  template <typename KVP>
+  constexpr RAFT_INLINE_FUNCTION KVP operator()(const KVP& a, const KVP& b) const
+  {
+    if ((b.value < a.value) || ((a.value == b.value) && (b.key < a.key))) { return b; }
+    return a;
+  }
+};
+
+struct argmax_op {
+  template <typename KVP>
+  constexpr RAFT_INLINE_FUNCTION KVP operator()(const KVP& a, const KVP& b) const
+  {
+    if ((b.value > a.value) || ((a.value == b.value) && (b.key < a.key))) { return b; }
+    return a;
+  }
+};
+
+template <typename ScalarT>
+struct const_op {
+  const ScalarT scalar;
+
+  constexpr const_op(ScalarT s) : scalar{s} {}
+
+  template <typename InT>
+  constexpr RAFT_INLINE_FUNCTION ScalarT operator()(InT unused) const
+  {
+    return scalar;
+  }
+};
+
+template <typename ComposedOpT, typename ScalarT>
+struct scalar_op {
+  ComposedOpT composed_op;
+  const ScalarT scalar;
+
+  constexpr scalar_op(ScalarT s) : scalar{s} {}
+
+  template <typename InT>
+  constexpr RAFT_INLINE_FUNCTION auto operator()(InT a) const
+  {
+    return composed_op(a, scalar);
+  }
+};
+
+template <typename Type>
+using scalar_add_op = scalar_op<add_op, Type>;
+
+template <typename Type>
+using scalar_sub_op = scalar_op<sub_op, Type>;
+
+template <typename Type>
+using scalar_mul_op = scalar_op<mul_op, Type>;
+
+template <typename Type>
+using scalar_div_op = scalar_op<div_op, Type>;
+
+template <typename Type>
+using scalar_div_checkzero_op = scalar_op<div_checkzero_op, Type>;
+
+template <typename Type>
+using scalar_pow_op = scalar_op<pow_op, Type>;
+
+template <typename OuterOpT, typename InnerOpT>
+struct unary_compose_op {
```

Comment:
I'm ok with this name `unary_compose_op`, but maybe it could also make sense to change it to just `compose_op`? My intuition here is that (1) users may mistakenly think that the `inner_op` is limited to one argument (2) it may be used all over the place, so shorter name could pay off :)